app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java

```diff
@@ -0,0 +1,355 @@
+                        Manifest.permission.ACCESS_FINE_LOCATION);
+
+        // Provide an additional rationale to the user. This would happen if the user denied the
+        // request previously, but didn't check the "Don't ask again" checkbox.
+        if (shouldProvideRationale) {
+            Log.i(TAG, "Displaying permission rationale to provide additional context.");
+            showSnackbar(R.string.permission_rationale, android.R.string.ok,
+                    new View.OnClickListener() {
+                        @Override
+                        public void onClick(View view) {
+                            // Request permission
+                            ActivityCompat.requestPermissions(mActivity,
+                                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
+                                    ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE);
+                        }
+                    });
+        } else {
+            Log.i(TAG, "Requesting permission");
+            // Request permission. It's possible this can be auto answered if device policy
+            // sets the permission in a given state or the user denied the permission
+            // previously and checked "Never ask again".
+            ActivityCompat.requestPermissions(mActivity,
+                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
+                    ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE);
+        }
+    }
+
+    /**
+     * Callback received when a permissions request has been completed. Redirected from activity.
+     */
+    public void onRequestPermissionsResult(@NonNull int[] grantResults) {
+        if (grantResults.length <= 0) {
+            // If user interaction was interrupted, the permission request is cancelled and you
+            // receive empty arrays.
+            Log.i(TAG, "User interaction was cancelled.");
+        } else if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            Log.i(TAG, "Permission granted.");
+            performPendingGeofenceTask();
+        } else {
+            // Permission denied.
+
+            // Notify the user via a SnackBar that they have rejected a core permission for the
+            // app, which makes the Activity useless. In a real app, core permissions would
+            // typically be best requested during a welcome-screen flow.
+
+            // Additionally, it is important to remember that a permission might have been
+            // rejected without asking the user for permission (device policy or "Never ask
+            // again" prompts). Therefore, a user interface affordance is typically implemented
+            // when permissions are denied. Otherwise, your app could appear unresponsive to
+            // touches or interactions which have required permissions.
+            showSnackbar(R.string.permission_denied_explanation, R.string.settings,
```

Comment:
The best practice is to ask for permissions only when they are needed while letting the user use all features of the app that don't require the permission. 
Exiting the app if permission not granted is horrible experience...
Because this is a core permission to our app it is possible to maybe ask for permission also in the beginning of the app but if not granted just go on until it is needed and ask again.